### PR TITLE
feat(dashboard): added new `dashboard` snack

### PR DESCRIPTION
**Pull request by folke**
## Description

A new dashboard plugin that comes with sane defaults, and can be fully customized to your liking.

## Screenshots

image

image

image

image

image

### CODE REVIEW

File: doc/snacks-dashboard.txt [hunk 1]
```diff
@@ -110,33 +110,22 @@ In the example below, both sections are equivalent.
 
 >lua
     ---@class snacks.dashboard.Config
-    ---@field sections (snacks.dashboard.Section|fun(opts:snacks.dashboard.Opts):(snacks.dashboard.Section|snacks.dashboard.Section[]|nil))[]
-    ---@field formats table<string, snacks.dashboard.Text|fun(value:string):snacks.dashboard.Text>
+    ---@field sections snacks.dashboard.Section
+    ---@field formats table<string, snacks.dashboard.Text|fun(item:snacks.dashboard.Item, ctx:sncaks.dashboard.Format.ctx):snacks.dashboard.Text>
```
Comment (dpetka2001) [inline doc/snacks-dashboard.txt:114]: Minor typo: `ctx:snacks.dashboard.Format.ctx` here and the other files as well?

Comment (folke) [reply doc/snacks-dashboard.txt]: Yep indeed. Just pushed a big refactor :)

File: doc/snacks-dashboard.txt [hunk 2]
```diff
@@ -12,8 +12,8 @@ Table of Contents                         *snacks-dashboard-table-of-contents*
 4. Types                                              |snacks-dashboard-types|
 5. Module                                            |snacks-dashboard-module|
   - Snacks.dashboard()            |snacks-dashboard-module-snacks.dashboard()|
+  - Snacks.dashboard.file_icon()|snacks-dashboard-module-snacks.dashboard.file_icon()|
   - Snacks.dashboard.have_pugin()|snacks-dashboard-module-snacks.dashboard.have_pugin()|
```
Comment (dpetka2001) [inline doc/snacks-dashboard.txt:16]: Sorry again. `have_plugin` instead? Here and the other files. `dashboard.lua` also has it misspelled not in this diff, but I looked directly at the file to make sure.

**Comment by folke** [comment]
Added a bunch of changes including terminal sections.
The terminal output is cached in a file for super fast rendering.
So only the first time it's slower.

**Comment by folke** [comment]
Like in the updated screenshots, where you can use a teminal command to generate output to show on your dashboard.

alpha.nvim and others have this too, but the command executes every time.

With what I just added, it only executes the first time, after that it renders the terminal output instantly.

**Comment by folke** [comment]
Just added the option to have multiple vertical `panes` to layout sections.
(code not pushed yet)

image

**Comment by dpetka2001** [comment]
This looks beautiful :)

**Comment by folke** [comment]
This is how the config looks like:

```lua
{
    { section = "header", enabled = true },
    {
      section = "terminal",
      pane = 2,
      cmd = "colorscript -e square",
      height = 5,
      padding = 1,
    },
    {
      section = "keys",
      gap = 1,
      padding = 1,
      enabled = true,
    },
    {
      pane = 2,
      title = "Recent Files",
      icon = " ",
      section = "recent_files",
      indent = 2,
      padding = 1,
    },
    {
      pane = 2,
      title = "Projects",
      icon = " ",
      indent = 2,
      section = "projects",
      limit = 10,
      padding = 1,
    },
    {
      title = "Git Status",
      icon = " ",
      section = "terminal",
      pane = 2,
      cmd = "hub st",
      height = 5,
      padding = 1,
      ttl = 5 * 60,
      indent = 3,
    },
    { section = "startup" },
  }
```

**Comment by dpetka2001** [comment]
Can the header also accept a `cmd` like the terminal section? I was playing around with `figlet` and `lolcat` and tried to create a `Lazyvim` logo with colors. I then saved that into a txt file and read somewhere (but did not actually try it yet) that you can use that in `cmd = cat logo.txt` to get a colorful logo in Neovim. Or we should use instead of a header section a terminal section for this?

File: lua/snacks/dashboard.lua [hunk 3]
```diff
@@ -489,49 +525,104 @@ function D:on(event, cb)
...
+---@param from? {[1]:number, [2]:number}
+function D:find(pos, from)
+  from = from or pos
+  local line = vim.api.nvim_buf_get_lines(self.buf, pos[1] - 1, pos[1], false)[1] or ""
+  local char = vim.fn.charidx(line, pos[2]) -- map col to charachter index
+
+  local pane = math.floor((char - self.col) / (self.opts.width + self.opts.pane_gap)) + 1
+  pane = math.max(1, math.min(pane, self.panes))
+  if pos[1] == from[1] then
+    dd({ pos = pos, from = from, pane = pane })
```
Comment (dpetka2001) [inline lua/snacks/dashboard.lua:538]: I'm guessing this is from your debug?

**Comment by folke** [comment]
The `header` section is just a simple text section that includes `opts.preset.header`.
You can instead just use a terminal section as the header.
You can use that `cat` approach, but with my impl it doesn't really matter, so you can also just use lolcat and others directly.

What happens under the hood is that I execute the cmd in a tty and then stream the output to a file which gets read and rendered using `nvim_open_term` and `nvim_chan_send`, without the need for any executing commands.

**Comment by folke** [comment]
So remove the `header` section and replace it with a terminal section and a cmd like `cmd = "figlet Neovim | lolcat"`

**Comment by dpetka2001** [comment]
I will wait until you fully merge this and play around then :smile: 

`lolcat` outputs different colors each time you execute it. So, something like this `figlet -f ~/.config/LazyVimDev/ansi_shadow.flf -c "LAZYVIM" | lolcat -F 0.7 -t -p 10` outputs different colors each time it executes. 

So, the first reason for using `cat` was to pipe the result to a txt file until I find a combination that I like :stuck_out_tongue: 
The second reason was that because I don't understand how the caching for the terminal command actually works. With the above code snippet, will it only execute once the first time and then just use whatever the first output of `lolcat` is for the next times? Or because `lolcat` outputs different colors, it will always execute like it's the first time?

**Comment by folke** [comment]
Since it will be cached, it will always show one variation indeed, but that's what the `random = 10` option is for.
Setting it to `10` would cache `10` random variations :)

**Comment by dpetka2001** [comment]
Love it!! Thanks a lot :smiley: 



Went the compact way on my 15.1" laptop :stuck_out_tongue:

**Comment by folke** [comment]
Nice!
With a shiny lolcat logo and all :)

**Comment by oktal** [comment]
Hey, that's very cool !

I'd like to try it out with lazyvim but can't get it working. Here's my config

```lua
return {
  { "nvimdev/dashboard-nvim", enabled = false },
  {
    "folke/snacks.nvim",
    opts = {
      dashboard = {
        enabled = true,
      },
    },
  },
}
```

Dashboard does not show up when starting LazyVim

`snacks` is loaded

```
    ● snacks.nvim 1.44ms  start
        dir    /Users/oktal/.local/share/nvim/lazy/snacks.nvim
        url    https://github.com/folke/snacks.nvim
        branch main
        commit c8e209e
        readme README.md
        help   |snacks-bigfile.txt|
        help   |snacks-bufdelete.txt|
        help   |snacks-dashboard.txt|
        help   |snacks-debug.txt|
        help   |snacks-git.txt|
        help   |snacks-gitbrowse.txt|
        help   |snacks-init.txt|
        help   |snacks-lazygit.txt|
        help   |snacks-notifier.txt|
        help   |snacks-notify.txt|
        help   |snacks-quickfile.txt|
        help   |snacks-rename.txt|
        help   |snacks-statuscolumn.txt|
        help   |snacks-terminal.txt|
        help   |snacks-toggle.txt|
        help   |snacks-util.txt|
        help   |snacks-win.txt|
        help   |snacks-words.txt|
        help   |snacks.nvim.txt|
        keys    <leader>un 

      ✔ [task] fetch 1009.95ms
      ✔ [task] status 23.05ms
      ✔ [task] log 111.2ms
```

Any idea ?

**Comment by folke** [comment]
@oktal what does checkhealth snacks show?

**Comment by dpetka2001** [comment]
Any idea how I can change the key for Lazy in the `preset.keys`?

**Comment by oktal** [comment]
> @oktal what does checkhealth snacks show?

```
snacks: require("snacks.health").check()

Snacks ~
- OK setup called
- OK `Snacks.bigfile` setup {enabled}
- OK `Snacks.lazygit` {lazygit} installed
- OK `Snacks.notifier` setup {enabled}
- OK `Snacks.notifier` is ready
- OK `Snacks.quickfile` setup {enabled}
- WARNING `Snacks.statuscolumn` setup {disabled}
- OK `Snacks.statuscolumn` is manually configured
  - `vim.o.statuscolumn = "%!v:lua.require'snacks.statuscolumn'.get()"`
- OK `Snacks.terminal` shell configured
  - `vim.o.shell`: /bin/zsh
  - `parsed`: { "/bin/zsh" }
- OK `Snacks.toggle` {which-key} is installed
- OK `Snacks.words` setup {enabled}
```

**Comment by dpetka2001** [comment]
@octal You did update `snacks.nvim` to latest version right? Just checking.

**Comment by dpetka2001** [comment]
So, I ended up passing the whole `dashboard.preset.keys` and just modifying the key for Lazy. It is a list after all and if I just added my modification it would overwrite it and only the new mapping for Lazy would appear. 

@folke If you know of a more proper way to do that I would appreciate any hints.

**Comment by folke** [comment]
Just pushed an update where you can now do something like:

```lua
        preset = {
          keys = function(keys)
            for _, key in ipairs(keys) do
              if key.key == "L" then
                key.key = "l"
              end
            end
          end,
        },

```

**Comment by dpetka2001** [comment]
Yep, it works perfect!! Tnx again :smile:

**Comment by folke** [comment]
These things now also work :) `cmatrix` and `tty-clock -s` as examples

https://github.com/user-attachments/assets/042ee5fe-f464-426f-9be8-3cfbd905fe7d

**Comment by LV** [comment]
@folke In the documentation, I'd add some sort of reference to `colorscripts` or `pokemon-colorscripts`. I assume these are the scripts in question: [1](https://gitlab.com/dwt1/shell-color-scripts), [2](https://github.com/nuke-dash/pokemon-colorscripts-mac).

I don't think it's uncommon for people to read the docs and want to replicate what you have as an example. It wouldn't hurt in my opinion.

**Pull request merged**